lib/Internal/List.ml

```diff
@@ -0,0 +1,16 @@
+include Stdlib.List
+
+(** [singleton element] creates a list with a single [element]. *)
+let singleton : 'a -> 'a t = fun element -> [ element ]
+
+(** [intersperse item list] inserts [item] between every
+      element of [list]. This function is tail-recursive. *)
+let intersperse : 'a -> 'a t -> 'a t =
+  let rec tailrec item list acc =
+    match list with
+    | [] -> rev acc
+    | last :: [] -> tailrec item [] (last :: acc)
+    | first :: rest -> tailrec item rest (item :: first :: acc)
```

Comment:
I never tried it but I wonder if the `[@tail_mod_cons]` annotation will be able to handle the last case, so you won't need to `rev` at the end:

- https://ocaml.org/manual/5.1/tail_mod_cons.html

```suggestion
  let[@tail_mod_cons] rec tailrec item list =
    match list with
    | [] -> []
    | [last] -> [last]
    | first :: rest -> first :: item :: tailrec item rest
```

If this doesn't work out of the box, then I still feel it should be possible to use a pair of mutually recursive functions with `[@tail_mod_cons]`. But at this point, I'm not sure which one will be more complex.

Reply:
Thank you for the suggestion, I will try it out!